### PR TITLE
Add ParameterizedType

### PR DESCRIPTION
**Pull request by TheFirstAvenger**
This is v1 of `ParameterizedType` as outlined in [this comment](https://github.com/elixir-ecto/ecto/pull/3339#issuecomment-670081678) on a previous iteration of this work. `ParameterizedType` is added, and `Ecto.Type.Enum` is added in `test` as the validation implementation of ParameterizedType (to be promoted to `lib` in a follow up PR).

@josevalim I modified the Embedded changes in the previous commit slightly, as all of the ParameterizedType callbacks have the options/config as the final argument, not the first (making them read left-to-right closer to the Ecto.Type callbacks).

### CODE REVIEW

File: lib/ecto/parameterized_type.ex [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      use Ecto.ParameterizedType
+
+      def type(_opts), do: :string
+
+      def init(opts) do
+        ...
+        Enum.into(opts, %{})
+      end
+
+      def cast(data, current, opts) do
```
Comment (josevalim) [inline lib/ecto/parameterized_type.ex:22]: Instead of calling the third element `opts`, let's call it `config` or something else. `opts` kind of implies keyword. :)

File: lib/ecto/parameterized_type.ex [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    schema "foo" do
+      field "bar", {MyApp.MyType, opt1: :baz, opt2: :boo}
+    end
+
+  """
+
+  @type opts :: keyword() | map()
+
+  @callback init(opts :: keyword()) :: opts()
```
Comment (josevalim) [inline lib/ecto/parameterized_type.ex:52]: I believe the field name would also be given on init, correct?

File: lib/ecto/parameterized_type.ex [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  @callback equal?(value :: any(), value :: any(), opts :: opts()) :: boolean()
+
+  @callback embed_as(format :: atom(), opts :: opts()) :: :self | :dump
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Ecto.ParameterizedType
+      def load(term, _loader, opts), do: load(term, opts)
+      def dump(term, _dumper, opts), do: dump(term, opts)
```
Comment (josevalim) [inline lib/ecto/parameterized_type.ex:76]: We definitely don't need both `load/2` and `load/3`, so my suggestion is to keep it simply as `load/3` and `dump/3` and simplify the API.

File: lib/ecto/parameterized_type.ex [hunk 4]
```diff
@@ -0,0 +1,83 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
```
Comment (josevalim) [inline lib/ecto/parameterized_type.ex:7]: We should also a paragraph that explains that, opposite to regular types, custom types are responsible for handling `nil` in both `load` and `dump` callbacks.

File: test/ecto/type/enum_test.exs [hunk 5]
```diff
@@ -0,0 +1,98 @@
+defmodule Ecto.Type.EnumTest do
```
Comment (josevalim) [inline test/ecto/type/enum_test.exs:1]: I am thinking this test is a bit too high-level for now. It probably makes sense to unit test those through the Ecto.Type API. We also need tests for equal?.

Comment (josevalim) [inline test/ecto/type/enum_test.exs:1]: Those tests are good but they are too high-level. We also need to unit test parameterized types directly through the Ecto.Type API.

File: lib/ecto/parameterized_type.ex (hunk 2 above)
Comment (josevalim) [reply lib/ecto/parameterized_type.ex:52]: Oh, I see you are passing it as an option. I am thinking it may be better to make this become: `init(schema, field, value)` and we don't put any restriction on the value. Thoughts?

File: lib/ecto/type.ex [hunk 6]
```diff
@@ -714,19 +718,24 @@ defmodule Ecto.Type do
       :error
 
   """
-  @spec cast(t, term) :: {:ok, term} | {:error, keyword()} | :error
-  def cast({:parameterized, module, config}, value), do: module.cast(config, value)
-  def cast({:in, _type}, nil), do: :error
-  def cast(_type, nil), do: {:ok, nil}
+  @spec cast(t, term, term) :: {:ok, term} | {:error, keyword() | Ecto.Changeset.t()} | :error | :ignore
+  def cast(_, _, current \\ nil)
```
Comment (josevalim) [inline lib/ecto/type.ex:722]: So this is a bit complicated... `cast` is called both by `Ecto.Changeset` and also by `Ecto.Query`. Once we pass `current` here, we are lumping both of them together, but in one of them we will never actually have a `current`. To make matters further apart, an embed would never really cast for a query, only for a changeset.

So it may be that, for parameterized types, we will have two cast functions, one receives the current value and the other does not. We can also give different names to them. 

For `cast+current`, we don't need to add it to Ecto.Type, we can call it directly from the changeset. So in a nutshell, my suggestion is to remove the extra `current` argument from here for now. We will add a new `cast` function once we work on the changeset.

File: lib/ecto/parameterized_type.ex (hunk 2 above)
Comment (Eiji7) [reply lib/ecto/parameterized_type.ex:52]: Sorry, maybe I do not see something, but what is the purpose of passing `schema` and `field` to any callback? I don't see a use case and therefore I don't believe anybody would need those. Also anything dynamic like `enum` values should be passed to `attrs`/`config`/`opts`/or however we name it, right?

**Comment by josevalim** [comment]
Awesome job @TheFirstAvenger, I have added some comments. I did not comment too much on the Enum module for now - but since it does touch some of the Ecto.Schema API, I would like to discuss the API for specifying parameterized modules.

I am not a big fan of the current approach where `{Mod, opts}` get translated to `{:parameterized, Mod, opts}`. If we want to go the translation approach, then I would rather do this:

```elixir
field :foo, Ecto.Enum, values: [:foo, :bar, :baz]
```

Then we basically need to check if the type is a parameterized type or basic type based on its implementation. The benefit of this approach is that for a user, there is no distinction between a parameterized API and a regular one, so an implementation can even swap between them without breaking user compatibility.

However, if we don't want to have translation, then I would rather have users specify the whole thing:

```elixir
field :foo, {:parameterized, Ecto.Enum, values: [:foo, :bar, :baz]}
```

It is a bit too verbose, so users could wrap it as:

```elixir
field :foo, enum([:foo, :bar, :baz])
```

But this feels like pointless indirection.

Thoughts?

File: lib/ecto/parameterized_type.ex (hunk 2 above)
Comment (josevalim) [reply lib/ecto/parameterized_type.ex:52]: @Eiji7 enum definitely needs them, but we get to cheat because we have a macro for passing them in. So I can see others needing it in the future. The whole point of `init` is to receive all information so you decide what to keep. What you keep is what becomes `config`.

**Comment by Eiji7** [comment]
> ```elixir
> field :foo, Ecto.Enum, values: [:foo, :bar, :baz]
> ```

I thought that there is already decision about API, but I was wrong. I really like this idea - that looks definitely better than previous one.

**Comment by TheFirstAvenger** [comment]
> ```elixir
> field :foo, Ecto.Enum, values: [:foo, :bar, :baz]
> ```

Yes, I much prefer this. I had similarly thought the decision was to go with the 2-tuple, but I will update it to this much cleaner format.

File: lib/ecto/parameterized_type.ex (hunk 1 above)
Comment (TheFirstAvenger) [reply lib/ecto/parameterized_type.ex:22]: @josevalim Good point. It initially was always a keyword. Now that the standard will be to convert it to a map in init, opts doesn't work. How about `params` since it is a ParameterizedType?

Comment (josevalim) [reply lib/ecto/parameterized_type.ex:22]: To me the params are the arguments given on `init` and not the inner state itself. That said, I am also fine with `params`. Anything besides `opts` is game for me. :)

File: lib/ecto/parameterized_type.ex (hunk 2 above)
Comment (josevalim) [reply lib/ecto/parameterized_type.ex:52]: Btw, if we decide to go with `field :foo, Ecto.Enum, values: [...]`, then I think this signature is perfect, as it will always be a keyword list. We can also pass the `:schema` and `:field` as an option too, the same as today. So I would just keep it as is. You can consider this particular comment resolved. :)

File: lib/ecto/schema.ex [hunk 7]
```diff
@@ -1771,11 +1771,35 @@ defmodule Ecto.Schema do
 
   @doc false
   def __field__(mod, name, type, opts) do
+    type = maybe_to_parameterized(type, name, mod)
```
Comment (josevalim) [inline lib/ecto/schema.ex:1774]: Instead of doing an early pass, change `check_field_type!` to return the type and we do all traversal there. Especially if we are going with `field :foo, Ecto.Enum, values: [...]`, since `check_field_type!` already has logic to validate custom types.

File: lib/ecto/parameterized_type.ex (hunk 1 above)
Comment (TheFirstAvenger) [reply lib/ecto/parameterized_type.ex:22]: Sounds good. I am thinking that `opts` will remain the name of the keyword that is passed into init (including what we call it in schema as we are parsing the field macro), and params is the map that is returned from init, and everywhere that is used in subsequent callbacks. As an example, ParameterizedType would have these:

```elixir
  @type opts :: keyword()

  @type params :: map()

  @callback init(opts :: opts()) :: params()

  @callback type(params :: params()) :: Ecto.Type.t()
```

File: lib/ecto/type.ex (hunk 6 above)
Comment (TheFirstAvenger) [reply lib/ecto/type.ex:722]: Sounds good. Yeah, `current` came into play with `Embed`, but isn't necessary in v1.

File: lib/ecto/parameterized_type.ex (hunk 1 above)
Comment (josevalim) [reply lib/ecto/parameterized_type.ex:22]: Params should be any `term`. There is no reason for us to impose it being a map. :)

**Comment by TheFirstAvenger** [comment]
@josevalim I have implemented all of the PR feedback, with the exception of adding tests specifically for ParameterizedType, which I will tackle tomorrow. Let me know if there is anything else missing or incorrect.

File: lib/ecto/parameterized_type.ex [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  @callback embed_as(format :: atom(), params :: params()) :: :self | :dump
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Ecto.ParameterizedType
+      def embed_as(_, _), do: :self
+      def equal?(term1, term2, _params), do: term1 == term2
+      def init(opts), do: Enum.into(opts, %{})
```
Comment (josevalim) [inline lib/ecto/parameterized_type.ex:78]: Let's remove the default implementation for `init`. I don't think it is a good idea for the default to simply collecting everything. Typically, you will want to validate options and keep only what you need. You will also want to precompute as much as you can to avoid compile time work. For example, for enums, you would do something like this:

```elixir
def init(opts) do
  values = Keyword.fetch!(opts, :values)
  user_to_db = Map.new(values, fn k -> {k, Atom.to_string(v)} end)
  db_to_user = Map.new(values, fn k -> {Atom.to_string(v), k} end)
  %{user_to_db: user_to_db, db_to_user: db_to_user}
end
```

So load and dump happens as efficiently as possible. For `Ecto.Embedded`, you can make it so:

```elixir
def init(_opts), do: raise "TODO: invoke this directly"
```

File: lib/ecto/parameterized_type.ex [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  @callback equal?(value1 :: any(), value2 :: any(), params :: params()) :: boolean()
+
+  @callback embed_as(format :: atom(), params :: params()) :: :self | :dump
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Ecto.ParameterizedType
+      def embed_as(_, _), do: :self
+      def equal?(term1, term2, _params), do: term1 == term2
```
Comment (josevalim) [inline lib/ecto/parameterized_type.ex:77]: I have a hunch we won't need `equal?/3` once we add `cast/3` and `change/3` because those callbacks can simply return `ignore` instead. So can you please add a TODO?

```suggestion
      # TODO: evaluate if we should keep this once we add cast/3 and change/3
      def equal?(term1, term2, _params), do: term1 == term2
```

File: lib/ecto/repo/schema.ex [hunk 10]
```diff
@@ -899,9 +899,13 @@ defmodule Ecto.Repo.Schema do
       {:ok, value} ->
         value
       :error ->
+        type_str = case type do
+          {:parameterized, t, _} -> t
+          _ -> type
+        end
```
Comment (josevalim) [inline lib/ecto/repo/schema.ex:905]: I would revert this change simply because in all other errors it will show up as `{:paramterized, ..., ...}`, so I wouldn't sweat about it for now.

File: lib/ecto/schema.ex [hunk 11]
```diff
@@ -1771,11 +1771,15 @@ defmodule Ecto.Schema do
...
   @doc false
   def __field__(mod, name, type, opts) do
-    check_field_type!(name, type, type, opts)
+    # type = convert(type)
+    type = check_field_type!(mod, name, type, type, opts)
     Module.put_attribute(mod, :changeset_fields, {name, type})
     define_field(mod, name, type, opts)
   end
 
+  # defp convert(:enum), do: Ecto.Type.Enum
```
Comment (josevalim) [inline lib/ecto/schema.ex:1780]: Btw, in the future we will simply call it `Ecto.Enum` and we won't need to convert this. We already do this for `Ecto.UUID`, so it is fine. :) So you can remove the commented out code!

File: lib/ecto/parameterized_type.ex (hunk 8 above)
Comment (TheFirstAvenger) [reply lib/ecto/parameterized_type.ex]: Done. I really like this mapping method, it makes cast/load/dump much nicer as declarative instead of imperative.

File: lib/ecto/schema.ex (hunk 11 above)
Comment (TheFirstAvenger) [reply lib/ecto/schema.ex]: Done. `Ecto.Type.Enum` and `:enum` are both now `Ecto.Enum`

**Comment by TheFirstAvenger** [comment]
@josevalim All feedback addressed and pushed, starting on ParameterizedType tests now.

**Comment by TheFirstAvenger** [comment]
@josevalim ParameterizedTypeTest added. Took a bit longer because in comparing with Ecto.Type tests, I realized that we haven't addressed :parameterized in :array, :map, or :maybe composite types. I wrote tests which I believe reflect how it _should_ work, however they don't currently work, and my initial pass at trying to get them to work got messy, so I pulled back to just testing what we have currently working, and added `:skip` to the :array, :map, and :maybe tests.

File: lib/ecto/type.ex [hunk 12]
```diff
@@ -439,6 +439,10 @@ defmodule Ecto.Type do
     array(value, type, dumper, [])
   end
 
+  def dump({:parameterized, type, params}, value, dumper) do
+    type.dump(value, dumper, params)
+  end
+
```
Comment (josevalim) [inline lib/ecto/type.ex:445]: I think we can remove this. It is a duplicate of the same clause on line 412:

```suggestion
```

File: test/ecto/parameterized_type_test.exs [hunk 13]
```diff
@@ -0,0 +1,143 @@
+defmodule Ecto.ParameterizedTypeTest do
+  use ExUnit.Case, async: true
+
+
```
Comment (josevalim) [inline test/ecto/parameterized_type_test.exs:4]:
```suggestion
```

File: test/ecto/type/enum_test.exs [hunk 14]
```diff
@@ -0,0 +1,120 @@
+defmodule Ecto.EnumTest do
+  use ExUnit.Case
```
Comment (josevalim) [inline test/ecto/type/enum_test.exs:2]:
```suggestion
  use ExUnit.Case, async: true
```

Also please rename the test file to `test/ecto/enum_test.exs`.

File: test/ecto/parameterized_type_test.exs [hunk 15]
```diff
@@ -0,0 +1,143 @@
+defmodule Ecto.ParameterizedTypeTest do
```
Comment (josevalim) [inline test/ecto/parameterized_type_test.exs:1]: Great job!

File: test/support/enum.ex [hunk 16]
```diff
@@ -0,0 +1,64 @@
+defmodule Ecto.Enum do
```
Comment (josevalim) [inline test/support/enum.ex:1]: I think we can move this to lib/ecto/enum.ex to be quite honest. It is almost ready, we just really need docs.

File: test/support/enum.ex [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+      raise ArgumentError, "Ecto.Enum types must have a values option specified as a list of atoms, e.g. field :my_field, Ecto.Enum, values: [:foo, :bar]"
+    end
+
+    user_to_db = Map.new(values, &{&1, Atom.to_string(&1)})
+    db_to_user = Map.new(values, &{Atom.to_string(&1), &1})
+    %{user_to_db: user_to_db, db_to_user: db_to_user}
+  end
+
+  def values({:parameterized, __MODULE__, params}), do: values(params)
+  def values(%{user_to_db: user_to_db}), do: Map.keys(user_to_db)
```
Comment (josevalim) [inline test/support/enum.ex:29]: I don't think we should make this public. :)

File: test/support/enum.ex [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+  def load(data, _loader, %{db_to_user: db_to_user}) do
+    case db_to_user do
+      %{^data => as_atom} -> {:ok, as_atom}
+      _ -> :error
+    end
+  end
+
+  def dump(data, _dumper, %{user_to_db: user_to_db}) do
```
Comment (josevalim) [inline test/support/enum.ex:50]: We should probably handle `nil` explicitly here too, WDYT?

File: test/ecto/parameterized_type_test.exs [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    def params, do: @params
+
+    def init([some_opt: :some_opt_value, field: :my_type, schema: _]), do: @params
+    def type(@params), do: :custom
+    def load(_, _, @params), do: {:ok, :load}
+    def dump( _, _, @params),  do: {:ok, :dump}
+    def cast( _, @params),  do: {:ok, :cast}
+    def equal?(true, _, @params), do: true
+    def equal?(_, _, @params), do: false
+    def embed_as(_, @params), do: :dump
```
Comment (josevalim) [inline test/ecto/parameterized_type_test.exs:18]: Since `Ecto.Enum` has embed_as `:dump`, let's make this one `:self` so we can test different scenarios.

**Comment by josevalim** [comment]
@TheFirstAvenger I have dropped some comments. I will look into the failing tests tomorrow morning and hopefully we can merge this. Can you give me access to your fork just in case?

File: test/ecto/parameterized_type_test.exs (hunk 19 above)
Comment (TheFirstAvenger) [reply test/ecto/parameterized_type_test.exs:18]: Done. I also added tests for the error paths (I originally added it for :maybe, but that isn't working yet).

**Comment by TheFirstAvenger** [comment]
@josevalim Done, you have access. Heads up that I squashed all the commits and force-pushed to clean it up a bit.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by wojtekmach** [comment]
@TheFirstAvenger and others involved: Thank you, this is really exciting!

**Comment by josevalim** [comment]
I think we are good for the next step, which is to convert the callbacks in Relation - starting with the ones used by changesets - to Ecto.ParameterizedType. This is probably when we will introduce the specific `cast` and `change` callbacks. :)

**Comment by josevalim** [comment]
Oh, we also need to document the callbacks on "Ecto.ParamterizedType", right now none of them are documented.

File: lib/ecto/parameterized_type.ex [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+  For example, `field :foo, :string` behaves the same for every field.
+  On the other hand, `field :foo, Ecto.Enum, values: [:foo, :bar, :baz]`
+  will likely have a different set of values per field.
+
+  Note that options are specified as a keyword, but it is idiomatic to
+  convert them to maps inside `c:init/1` for easier pattern matching in
+  other callbacks.
+
+  Parameterized types are a superset of regualr types. In other words,
```
Comment (brianbowden) [inline lib/ecto/parameterized_type.ex:17]: Got a typo here! ("regualr")

**Comment by josevalim** [comment]
I have created an issue so we can continue the discussion of next steps: #3380 - i think the work in the previous PR will be relevant for us to move forward :)
